Add boosting k-NN exact search blog (#3448)
opensearch-project/project-website: _posts/2024-11-13-boosting-k-nn-exact-search.md

---
layout: post
title: "Boosting k-NN exact search performance"
authors:
   - ryanbogan
   - jmazane
   - vamshin
   - kolchfa

> date: 2024-11-13

Author: Done

> date: 2024-11-19

Reviewer: Please update date to: 2024-11-25

categories: 
    - technical-posts
has_science_table: true

> meta_keywords: exact k-NN search in OpenSearch, SIMD optimizations, query performance, relevance scoring, latency management

Author: ```suggestion
meta_keywords: k-NN search performance, SIMD in OpenSearch, script score queries, performance optimization, machine learning, exact K-NN
```

> meta_description: Learn how to achieve precise, high-performance search with exact k-NN in OpenSearch.

Reviewer: Please update the meta with the following:

**meta _keywords:** k-NN search performance, SIMD in OpenSearch, script score queries, performance optimization, machine learning, exact K-NN

**meta_description:** Boost exact k-NN search performance in OpenSearch using SIMD optimizations and script_score queries. Discover real-world performance gains for efficient vector similarity searches in machine learning applications.

Author: Updated

> meta_description: Learn how to achieve precise, high-performance search with exact k-NN in OpenSearch.

Author: ```suggestion
meta_description: Boost exact k-NN search performance in OpenSearch using SIMD optimizations and script_score queries. Discover real-world performance gains for efficient vector similarity searches in machine learning applications.
```

---

Exact k-nearest neighbor (k-NN) search in OpenSearch allows you to define custom scoring functions to retrieve documents based on their proximity to a query vector. This method provides highly accurate search results, making it ideal when you need precise, deterministic matches.

> Using OpenSearch's `script_score` queries, you can perform exact k-NN searches to find the closest neighbors to a query vector. This query type allows you to create complex scoring functions that account for factors like document attributes, user preferences, or external data. 

Reviewer: Add a link to here: https://opensearch.org/docs/latest/search-plugins/knn/knn-score-script/

{ "my_vector1": [1.5, 2.5], "price": 12.2 }
{ "index": { "_index": "my-knn-index-1", "_id": "2" } }
{ "my_vector1": [2.5, 3.5], "price": 7.1 }
// Additional documents omitted for brevity
```

Finally, run an exact k-NN search using the `script_score` query:

```json
GET my-knn-index-1/_search
{
 "size": 4,
 "query": {
   "script_score": {
     "query": { "match_all": {} },
     "script": {
       "source": "knn_score",
       "lang": "knn",
       "params": {
         "field": "my_vector2",
         "query_value": [2.0, 3.0, 5.0, 6.0],
         "space_type": "l2"
       }
     }
   }
 }
}
```

For more information about exact k-NN search and `script_score` queries, see [Exact k-NN with scoring script](https://opensearch.org/docs/latest/search-plugins/knn/knn-score-script/). You'll find detailed guides to help you configure k-NN exact search and make the most of custom scoring.

## Experiments show real-world performance gains

To measure the impact of these optimizations, we conducted A/B tests comparing OpenSearch 2.14 to OpenSearch 2.15 and 2.17 in a single-node cluster. 

### Cluster configuration

|Dataset	|Cohere 1m	|
|---	|---	|
|Data nodes	|1	|
|CPUs	|8	|

> |EBS Volume (GB)	|500	|

Reviewer 2: ```suggestion
|EBS volume (GB)	|500	|
```

### Results

> The following table provides latency comparison between OpenSearch versions 2.14 and 2.15.

Reviewer 2: ```suggestion
The following table provides a latency comparison between OpenSearch versions 2.14 and 2.15.
```

> |Space type	|Version	|50th percentile latency (ms)	|90th percentile latency (ms)	|99th percentile latency (ms)	|

Reviewer 3: Lets just look at 2.14 vs 2.17 and remove this table.

|---	|---	|---	|---	|---	|
|**Inner product**	|2.14	|668.84378	|816.95014	|948.21019	|
|| 2.15	|481.71112	|499.61605	|519.03619	|
|| **Improvement**	|27.98%	|38.84%	|45.26%	|
|**L2**	|2.14	|670.98628	|682.84925	|693.12135	|
|| 2.15	|503.96816	|520.86304	|537.51321	|
|| **Improvement**	|24.89%	|23.72%	|22.45%	|

> The following table provides latency comparison between OpenSearch versions 2.14 and 2.17.

Reviewer 2: ```suggestion
The following table provides a latency comparison between OpenSearch versions 2.14 and 2.17.
```

|Space type	|Version	|50th percentile latency (ms)	|90th percentile latency (ms)	|99th percentile latency (ms)	|
|---	|---	|---	|---	|---	|
|**Inner product**	|2.14	|668.84378	|816.95014	|948.21019	|
||2.17	|99.58072	|117.0792	|121.36626	|
||**Improvement**	|85.11%	|85.67%	|87.20%	|
|**L2**|2.14	|670.98628	|682.84925	|693.12135	|
| | 2.17	|104.36596	|118.85475	|127.60656	|
|| **Improvement**	|84.45%	|82.59%	|81.59%	|

### Conclusion

> The tests showed that OpenSearch's new SIMD optimizations resulted in significant latency reductions, especially for the inner product space type, which saw up to an 87% latency reduction at the 99th percentile.

Reviewer 3: Can we generalize SIMD to "new SIMD support and optimized memory access". We added optimized memory access in 2.17

Author: Done

## What's next for exact k-NN search?

> 

Reviewer 2: Line 123: In the second sentence, rather than "directly" preceding "at query time", should it instead precede "switch"?

Author: I think we can remove "directly". Let me do that.

> Future OpenSearch updates will provide even more flexibility for k-NN search. You'll be able to switch between exact and approximate search directly at query time. Additionally, future versions will provide the ability to specify which fields build indexes for exact and approximate search types. Stay tuned for these updates as we continue to improve OpenSearch's k-NN search capabilities.

Reviewer 2: ```suggestion
Future OpenSearch versions will provide even more k-NN search flexibility. You'll be able to switch between exact and approximate search directly at query time. Additionally, future versions will provide the ability to specify which fields build indexes for exact and approximate search types. Stay tuned for these updates as we continue to improve OpenSearch's k-NN search capabilities.
```
